<a href="https://colab.research.google.com/github/JBadawood/Kaggle/blob/main/Intro_to_ML-Kaggle.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a name='intro'></a>
# **Introduction to Machine Learning** 
*Learn the core ideas in machine learning, and build your first models.*\
https://www.kaggle.com/learn/intro-to-machine-learning

[GitHub](https://github.com/Kaggle/learntools/tree/master/notebooks/machine_learning)

**تعلم الآلة**\
*مجال فرعي للذكاء الاصطناعي **يهتم بتعلم الأنماط** من البيانات المتاحة لعمل **تنبؤات** أو **قرارات** مبنية على بيانات جديدة دون برمجة صريحة*

**Machine Learning (ML)**\
*An AI subfield concerned with **learning patterns** from available data to make **predictions** or **decisions** based on new data without explicit programming.*

>المصدر: [معجـــم البيانـــات والـــذكاء االصطناعـــي](https://sdaia.gov.sa/files/Dictionary.pdf) من الهيئة السعودية للبيانات والذكاء الاصطناعي بالتعاون مع مجمع الملك سلمان العالمي للغة العربية (الطبعة الأولى 2022)

____

**Kaggle**
> **Step1:** [Intro to Programming](https://www.kaggle.com/learn/intro-to-programming) (💯 DONE 🙂) \
**Step2:** [Python](https://www.kaggle.com/learn/python) (💯 DONE 🙂)\
**Step3:** [Intro to Machine Learning](https://www.kaggle.com/learn/intro-to-machine-learning) (NOW 💪)

> **[TensorFlow Guide](https://www.kaggle.com/learn-guide/tensorflow)**\
**Step4:** [Intro to Deep Learning](https://www.kaggle.com/learn/intro-to-deep-learning)\
**Step5:** [Computer Vision](https://www.kaggle.com/learn/computer-vision)

>**Step6:** [Pandas](https://www.kaggle.com/learn/pandas)\
**Step7:** [Intermediate Machine Learning](https://www.kaggle.com/learn/intermediate-machine-learning)\
**Step8:** [Feature Engineering](https://www.kaggle.com/learn/feature-engineering)\
**Step9:** [Time Series](https://www.kaggle.com/learn/time-series)

----
**Coursera | ML & DL**
>**Step10:** [Machine Learning Specialization](https://www.deeplearning.ai/program/machine-learning-specialization/) (DeepLearning.AI & Stanford)\
**Step11:** [The Deep Learning Specialization](https://www.deeplearning.ai/program/deep-learning-specialization/) (DeepLearning.AI & Stanford)

||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||

**Additional Resources**\
**Microsoft | AI for Beginners**
>- [Artificial Intelligence for Beginners](https://github.com/microsoft/ai-for-beginners)
>- [Classic Machine Learning for Beginners](https://github.com/microsoft/ML-For-Beginners)
>- [Data Science for Beginners](https://github.com/microsoft/Data-Science-For-Beginners)

**Machine Learning University (MLU) | MLU-EXPLAIN**
>- [Interactives/Visual explanations of core machine learning concepts](https://mlu-explain.github.io/)

==============================================================

**IBM & UDACITY | Data Science & Analysis**
> **Course1:** [Python for Data Science](https://cognitiveclass.ai/courses/python-for-data-science) == [AI Programming with Python](https://www.udacity.com/course/ai-programming-python-nanodegree--nd089)
\
**Course2:** [Data Analysis with Python](https://cognitiveclass.ai/courses/data-analysis-python)\
**Course3:** [Machine Learning with Python](https://cognitiveclass.ai/courses/machine-learning-with-python) or 
[Intro to Machine Learning with TensorFlow](https://www.udacity.com/course/intro-to-machine-learning-with-tensorflow-nanodegree--nd230)

# **Table of Contents**
[Intro. and Plan of Learning](#intro)

1. [How Models Work](#ch1)

2. [Basic Data Exploration](#ch2)\
  2.1 [Using Pandas to Get Familiar With Your Data](#sec2-1)\
  2.2 [Interpreting Data Description](#sec2-2)\
  [Exercise: Explore Your Data](#ex1)

3. [Your First Machine Learning Model](#ch3)\
  3.1 [Selecting Data for Modeling](#sec3-1)\
  3.2 [Choosing "Features"](#sec3-2)\
  3.3 [Building Your Model](#sec3-3)\
  [Exercise: Your First Machine Learning Model](#ex2)

4. [Model Validation](#ch4)\
  4.1 [What is Model Validation](#sec4-1)\
  4.2 [Calculate Mean Absolute Error (MAE)](#sec4-2)\
  4.3 [The Problem with "In-Sample" Scores](#sec4-3)\
  [Exercise: Your First Machine Learning Model](#ex3)

5. [Underfitting and Overfitting](#ch5)\
  5.1 [Compare MAE Scores](#sec5-1)\
  [Exercise: Underfitting and Overfitting](#ex4)

6. [Random Forests](#ch6)

7. [Machine Learning Competitions](#ch7)

[Bonus Exercise : Titanic Tutorial](#bonus)


**Install Kaggle exercises in Google Colab** ([link of steps](https://www.kaggle.com/questions-and-answers/165757))\
https://github.com/Kaggle/learntools

```
!git clone https://github.com/Kaggle/learntools.git
!mv learntools learntools_dir
!mv learntools_dir/learntools learntools
```



<a name='ch1'></a>
## **1- How Models Work**
The first step if you're new to machine learning.\
https://www.kaggle.com/code/dansbecker/how-models-work

قبل بناء أي نموذج نحتاج بيانات سابقة تم جمعها. مثلًا:

لو نبغى نتوقع أسعار البيوت الجديدة راح يتطلب مننا قبلها أننا نعرف كم كانت أسعار البيوت القديمة وكذلك الموجودة حاليًا (يعني نشوف الـ history)

بهذه الطريقة راح نعرف الـ patterns لهذه الأسعار السابقة ومن ثم يمكننا الاستفادة منها لعمل predictions للبيوت الجديدة

We'll start with a model called the **Decision Tree**. There are fancier models that give more accurate predictions. But decision trees are easy to understand, and they are *the basic building block for some of the best models* in **data science**.

> You can capture more factors using a tree that has more "**splits**." The most factors affecting home price, are: 
- number of bathrooms, 
- lot size, 
- location, etc.

> These are called **"deeper" trees**. The predicted price for the house is *at the bottom of the tree*. The point at the bottom where we make a prediction is called a **leaf**.

We use data to decide how to break the houses into two groups, and then again to determine the predicted price in each group. This step of capturing patterns from data is called **fitting** or **training** the model. The data used to **fit** the model is called the **training data**.

After the model has been fit, you can apply it to new data to **predict** prices of additional homes.

<a name='ch2'></a>
## **2- Basic Data Exploration**
Load and understand your data.\
https://www.kaggle.com/code/dansbecker/basic-data-exploration

<a name='sec2-1'></a>
### 2-1 Using Pandas to Get Familiar With Your Data
https://www.kaggle.com/code/dansbecker/basic-data-exploration?scriptVersionId=79125477&cellId=1

أول خطوة نحتاجها هي أننا نفهم البيانات اللي عندنا 

عشان كذا نحتاج pandas كوحدة من الأدوات اللي راح تساعدنا على استكشاف ومعالجة البيانات

In [ ]:
import pandas as pd

#### Melbourne Housing Snapshot
https://www.kaggle.com/datasets/dansbecker/melbourne-housing-snapshot

We'll look at data about home prices in Melbourne, Australia.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# راح يتم قراءة الملف وتخزينه في DataFrame
melbourne_data = pd.read_csv('/content/drive/MyDrive/iMLND/Kaggle/Intro_to_ML/melb_data.csv')

In [ ]:
# نقدر نشوف أول خمسة صفوف من هذا الملف
melbourne_data.head()

,Suburb,Address,Rooms,Type,Price,Method,SellerG,Date,Distance,Postcode,...,Bathroom,Car,Landsize,BuildingArea,YearBuilt,CouncilArea,Lattitude,Longtitude,Regionname,Propertycount
0,Abbotsford,85 Turner St,2,h,1480000.0,S,Biggin,3/12/2016,2.5,3067.0,...,1.0,1.0,202.0,NaN,NaN,Yarra,-37.7996,144.9984,Northern Metropolitan,4019.0
1,Abbotsford,25 Bloomburg St,2,h,1035000.0,S,Biggin,4/02/2016,2.5,3067.0,...,1.0,0.0,156.0,79.0,1900.0,Yarra,-37.8079,144.9934,Northern Metropolitan,4019.0
2,Abbotsford,5 Charles St,3,h,1465000.0,SP,Biggin,4/03/2017,2.5,3067.0,...,2.0,0.0,134.0,150.0,1900.0,Yarra,-37.8093,144.9944,Northern Metropolitan,4019.0
3,Abbotsford,40 Federation La,3,h,850000.0,PI,Biggin,4/03/2017,2.5,3067.0,...,2.0,1.0,94.0,NaN,NaN,Yarra,-37.7969,144.9969,Northern Metropolitan,4019.0
4,Abbotsford,55a Park St,4,h,1600000.0,VB,Nelson,4/06/2016,2.5,3067.0,...,1.0,2.0,120.0,142.0,2014.0,Yarra,-37.8072,144.9941,Northern Metropolitan,4019.0


In [ ]:
# هذا الكود عشان نشوف ملخص عن البيانات اللي عندنا
melbourne_data.describe()

,Rooms,Price,Distance,Postcode,Bedroom2,Bathroom,Car,Landsize,BuildingArea,YearBuilt,Lattitude,Longtitude,Propertycount
count,13580.000000,1.358000e+04,13580.000000,13580.000000,13580.000000,13580.000000,13518.000000,13580.000000,7130.000000,8205.000000,13580.000000,13580.000000,13580.000000
mean,2.937997,1.075684e+06,10.137776,3105.301915,2.914728,1.534242,1.610075,558.416127,151.967650,1964.684217,-37.809203,144.995216,7454.417378
std,0.955748,6.393107e+05,5.868725,90.676964,0.965921,0.691712,0.962634,3990.669241,541.014538,37.273762,0.079260,0.103916,4378.581772
min,1.000000,8.500000e+04,0.000000,3000.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1196.000000,-38.182550,144.431810,249.000000
25%,2.000000,6.500000e+05,6.100000,3044.000000,2.000000,1.000000,1.000000,177.000000,93.000000,1940.000000,-37.856822,144.929600,4380.000000
50%,3.000000,9.030000e+05,9.200000,3084.000000,3.000000,1.000000,2.000000,440.000000,126.000000,1970.000000,-37.802355,145.000100,6555.000000
75%,3.000000,1.330000e+06,13.000000,3148.000000,3.000000,2.000000,2.000000,651.000000,174.000000,1999.000000,-37.756400,145.058305,10331.000000
max,10.000000,9.000000e+06,48.100000,3977.000000,20.000000,8.000000,10.000000,433014.000000,44515.000000,2018.000000,-37.408530,145.526350,21650.000000


<a name='sec2-2'></a>
### 2-2 Interpreting Data Description
https://www.kaggle.com/code/dansbecker/basic-data-exploration?scriptVersionId=79125477&cellId=5

من ملخص البيانات السابق نقدر نعرف كل من:
- عدد الصفوف اللي ما فيها missing values (count)
- متوسط القيم حق العمود mean
- الإنحراف المعياري (عشان نعرف انتشار البيانات) (standard deviation)
- القيمة الصغرى min
- الربيعات Interquartile
> 1- الربيع الأدنى Q1(25%)\
2- الوسيط Q2(50%)\
3- الربيع الأعلى Q3(75%)
- القيمة العظمى max


<a name='ex1'></a>
### Exercise: Explore Your Data

In [ ]:
import pandas as pd

#### Step 1: Loading Data
Advance House Price Predicitons\
house prices from Iowa\
https://www.kaggle.com/datasets/zeeshanmulla/advance-house-price-predicitons

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# الملف اللي راح نقراه في الخطوة التالية
iowa_file_path = '/content/drive/MyDrive/iMLND/Kaggle/Intro_to_ML/house_prices_iowa/iowa/train.csv'

In [ ]:
# قراءة الملف عبر مكتبة pandas
home_data = pd.read_csv(iowa_file_path)

#### Step 2: Review The Data

In [ ]:
home_data.head()

,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.0,9550,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.0,14260,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,12,2008,WD,Normal,250000


In [ ]:
# Summary statistics
home_data.describe()

,Id,MSSubClass,LotFrontage,LotArea,OverallQual,OverallCond,YearBuilt,YearRemodAdd,MasVnrArea,BsmtFinSF1,...,WoodDeckSF,OpenPorchSF,EnclosedPorch,3SsnPorch,ScreenPorch,PoolArea,MiscVal,MoSold,YrSold,SalePrice
count,1460.000000,1460.000000,1201.000000,1460.000000,1460.000000,1460.000000,1460.000000,1460.000000,1452.000000,1460.000000,...,1460.000000,1460.000000,1460.000000,1460.000000,1460.000000,1460.000000,1460.000000,1460.000000,1460.000000,1460.000000
mean,730.500000,56.897260,70.049958,10516.828082,6.099315,5.575342,1971.267808,1984.865753,103.685262,443.639726,...,94.244521,46.660274,21.954110,3.409589,15.060959,2.758904,43.489041,6.321918,2007.815753,180921.195890
std,421.610009,42.300571,24.284752,9981.264932,1.382997,1.112799,30.202904,20.645407,181.066207,456.098091,...,125.338794,66.256028,61.119149,29.317331,55.757415,40.177307,496.123024,2.703626,1.328095,79442.502883
min,1.000000,20.000000,21.000000,1300.000000,1.000000,1.000000,1872.000000,1950.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,2006.000000,34900.000000
25%,365.750000,20.000000,59.000000,7553.500000,5.000000,5.000000,1954.000000,1967.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,5.000000,2007.000000,129975.000000
50%,730.500000,50.000000,69.000000,9478.500000,6.000000,5.000000,1973.000000,1994.000000,0.000000,383.500000,...,0.000000,25.000000,0.000000,0.000000,0.000000,0.000000,0.000000,6.000000,2008.000000,163000.000000
75%,1095.250000,70.000000,80.000000,11601.500000,7.000000,6.000000,2000.000000,2004.000000,166.000000,712.250000,...,168.000000,68.000000,0.000000,0.000000,0.000000,0.000000,0.000000,8.000000,2009.000000,214000.000000
max,1460.000000,190.000000,313.000000,215245.000000,10.000000,9.000000,2010.000000,2010.000000,1600.000000,5644.000000,...,857.000000,547.000000,552.000000,508.000000,480.000000,738.000000,15500.000000,12.000000,2010.000000,755000.000000


عشان نشوف كل الأعمدة نكتب الكود التالي:\
https://stackoverflow.com/questions/49188960/how-to-show-all-columns-names-on-a-large-pandas-dataframe

```
print(df.columns.tolist())
```



In [ ]:
print(home_data.columns.tolist())

['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating', 'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual', 'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond', 'PavedDrive', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'PoolQC'

In [ ]:
home_data.columns

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

In [ ]:
home_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

from data description file:

`LotArea`: *Lot size in square feet*

In [ ]:
home_data['LotArea'].mean()

10516.828082191782

In [ ]:
round(home_data['LotArea'].mean())

10517

In [ ]:
# What is the average lot size (rounded to nearest integer)?
avg_lot_size = round(home_data['LotArea'].mean())

print(f"The average lot size = {avg_lot_size}")


The average lot size = 10517


`YrSold`: *Year Sold (YYYY)*

In [ ]:
home_data['YrSold'].max()

2010

`YearBuilt`: Original construction date

In [ ]:
home_data.YearBuilt.max()

2010

In [ ]:
# As of today, how old is the newest home (current year - the date in which it was built)
newest_home_age = 2022 - home_data.YearBuilt.max()

print(f'The newest home age is {newest_home_age}')

The newest home age is 12


<a name='ch3'></a>
## **3- Your First Machine Learning Model**
Building your first model.\
https://www.kaggle.com/code/dansbecker/your-first-machine-learning-model

**[Melbourne Housing Snapshot](https://www.kaggle.com/datasets/dansbecker/melbourne-housing-snapshot)**


<a name='sec3-1'></a>
### 3-1 Selecting Data for Modeling
https://www.kaggle.com/code/dansbecker/your-first-machine-learning-model#Selecting-Data-for-Modeling

In [ ]:
import pandas as pd

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

In [ ]:
melbourne_data = pd.read_csv('/content/drive/MyDrive/iMLND/Kaggle/Intro_to_ML/melb_data.csv')

In [ ]:
# لرؤية جميع الـ variables
melbourne_data.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

> ايش الطريقة اللي عن طريقها راح نختار أفضل الـ variable واللي راح يكون لها تأثير كبير على دقة الـ model؟

We'll start by picking a few variables using our intuition. *Later courses will show you **statistical techniques to automatically prioritize variables**.*

In [ ]:
melbourne_data.head()

,Suburb,Address,Rooms,Type,Price,Method,SellerG,Date,Distance,Postcode,...,Bathroom,Car,Landsize,BuildingArea,YearBuilt,CouncilArea,Lattitude,Longtitude,Regionname,Propertycount
0,Abbotsford,85 Turner St,2,h,1480000.0,S,Biggin,3/12/2016,2.5,3067.0,...,1.0,1.0,202.0,NaN,NaN,Yarra,-37.7996,144.9984,Northern Metropolitan,4019.0
1,Abbotsford,25 Bloomburg St,2,h,1035000.0,S,Biggin,4/02/2016,2.5,3067.0,...,1.0,0.0,156.0,79.0,1900.0,Yarra,-37.8079,144.9934,Northern Metropolitan,4019.0
2,Abbotsford,5 Charles St,3,h,1465000.0,SP,Biggin,4/03/2017,2.5,3067.0,...,2.0,0.0,134.0,150.0,1900.0,Yarra,-37.8093,144.9944,Northern Metropolitan,4019.0
3,Abbotsford,40 Federation La,3,h,850000.0,PI,Biggin,4/03/2017,2.5,3067.0,...,2.0,1.0,94.0,NaN,NaN,Yarra,-37.7969,144.9969,Northern Metropolitan,4019.0
4,Abbotsford,55a Park St,4,h,1600000.0,VB,Nelson,4/06/2016,2.5,3067.0,...,1.0,2.0,120.0,142.0,2014.0,Yarra,-37.8072,144.9941,Northern Metropolitan,4019.0


In [ ]:
# عشان نشوف ايش الأعمدة اللي فيها null نكتب التالي
melbourne_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         13580 non-null  object 
 1   Address        13580 non-null  object 
 2   Rooms          13580 non-null  int64  
 3   Type           13580 non-null  object 
 4   Price          13580 non-null  float64
 5   Method         13580 non-null  object 
 6   SellerG        13580 non-null  object 
 7   Date           13580 non-null  object 
 8   Distance       13580 non-null  float64
 9   Postcode       13580 non-null  float64
 10  Bedroom2       13580 non-null  float64
 11  Bathroom       13580 non-null  float64
 12  Car            13518 non-null  float64
 13  Landsize       13580 non-null  float64
 14  BuildingArea   7130 non-null   float64
 15  YearBuilt      8205 non-null   float64
 16  CouncilArea    12211 non-null  object 
 17  Lattitude      13580 non-null  float64
 18  Longti

> نلاحظ أن عدد الصفوف هو 13580 بس بعض الأعمدة ما فيها قيم (فاضية) زي:
- BuildingArea
- YearBuilt 

In [ ]:
# راح نحذف الصفوف اللي فيها NA
melbourne_data = melbourne_data.dropna(axis=0)

In [ ]:
melbourne_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6196 entries, 1 to 12212
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         6196 non-null   object 
 1   Address        6196 non-null   object 
 2   Rooms          6196 non-null   int64  
 3   Type           6196 non-null   object 
 4   Price          6196 non-null   float64
 5   Method         6196 non-null   object 
 6   SellerG        6196 non-null   object 
 7   Date           6196 non-null   object 
 8   Distance       6196 non-null   float64
 9   Postcode       6196 non-null   float64
 10  Bedroom2       6196 non-null   float64
 11  Bathroom       6196 non-null   float64
 12  Car            6196 non-null   float64
 13  Landsize       6196 non-null   float64
 14  BuildingArea   6196 non-null   float64
 15  YearBuilt      6196 non-null   float64
 16  CouncilArea    6196 non-null   object 
 17  Lattitude      6196 non-null   float64
 18  Longtit

> بعد ما عملنا drop للـ missing values صارت عدد الصفوف 6196

هذه الطريقة مبدئية بس غير مجدية لأننا حذفنا قرابة نص البيانات

يعني قد تتسبب هذه الطريقة في أننا نحصل على Model غير دقيق

#### Selecting The Prediction Target

There are many ways to select a subset of your data. The [Pandas course](https://www.kaggle.com/learn/pandas) covers these in more depth, but we will focus on two approaches for now.

1. Dot notation, which we use to select the "**prediction target**"
2. Selecting with a column list, which we use to select the "**features**"

**dot-notation** : عن طريقها نقدر نخزن عمود واحد على أنه **Series**

> هذه الطريقة راح نختار بيها العمود اللي نبغى نتنبأ بالقيم حقه **prediction target**

In [ ]:
# The house prices (prediction target)
y = melbourne_data.Price

In [ ]:
y

1        1035000.0
2        1465000.0
4        1600000.0
6        1876000.0
7        1636000.0
           ...    
12205     601000.0
12206    1050000.0
12207     385000.0
12209     560000.0
12212    2450000.0
Name: Price, Length: 6196, dtype: float64

<a name='sec3-2'></a>
### 3-2 Choosing "Features"
https://www.kaggle.com/code/dansbecker/your-first-machine-learning-model#Choosing-%22Features%22

**Features:** columns that are inputted into model
> هي تمثل الأعمدة اللي راح نستخدمها لتحديد الـ home price

- قد نستخدم كل الأعمدة الموجودة في الجدول عدا العمود اللي استخدمناه في الـ target
- لكن في أوقات نحتاج نختار من هذه الأعمدة بدل ما نستخدمها كلها في الـ Model حقنا

الآن راح نستخدم بعض الـ features ونشوف كيف راح يكون لها تأثير على الـ model 

بعد كذا راح نستخدم features أخرى ونقارن الـ Models مع بعض

In [ ]:
melbourne_features = ['Rooms', 'Bathroom', 'Landsize', 'Lattitude','Longtitude']

In [ ]:
X = melbourne_data[melbourne_features]

In [ ]:
X

,Rooms,Bathroom,Landsize,Lattitude,Longtitude
1,2,1.0,156.0,-37.80790,144.99340
2,3,2.0,134.0,-37.80930,144.99440
4,4,1.0,120.0,-37.80720,144.99410
6,3,2.0,245.0,-37.80240,144.99930
7,2,1.0,256.0,-37.80600,144.99540
...,...,...,...,...,...
12205,3,2.0,972.0,-37.51232,145.13282
12206,3,1.0,179.0,-37.86558,144.90474
12207,1,1.0,0.0,-37.85588,144.89936
12209,2,1.0,0.0,-37.85581,144.99025


In [ ]:
# هنا عندنا ملخص للـ statistics حق الـ features
X.describe()

,Rooms,Bathroom,Landsize,Lattitude,Longtitude
count,6196.000000,6196.000000,6196.000000,6196.000000,6196.000000
mean,2.931407,1.576340,471.006940,-37.807904,144.990201
std,0.971079,0.711362,897.449881,0.075850,0.099165
min,1.000000,1.000000,0.000000,-38.164920,144.542370
25%,2.000000,1.000000,152.000000,-37.855438,144.926198
50%,3.000000,1.000000,373.000000,-37.802250,144.995800
75%,4.000000,2.000000,628.000000,-37.758200,145.052700
max,8.000000,8.000000,37000.000000,-37.457090,145.526350


In [ ]:
# راح نشوف بعض أول خمس صفوف من الـ feaures
X.head()

,Rooms,Bathroom,Landsize,Lattitude,Longtitude
1,2,1.0,156.0,-37.8079,144.9934
2,3,2.0,134.0,-37.8093,144.9944
4,4,1.0,120.0,-37.8072,144.9941
6,3,2.0,245.0,-37.8024,144.9993
7,2,1.0,256.0,-37.8060,144.9954


<a name='sec3-3'></a>
### 3-3 Building Your Model
https://www.kaggle.com/code/dansbecker/your-first-machine-learning-model#Building-Your-Model

لبناء نموذج تعلم الآلة راح نستخدم مكتبة scikit-learn (sklearn)

**Scikit-learn** is easily *the most popular library for modeling* the types of data typically stored in **DataFrames**.

The steps to building and using a model are:

- **Define**: What type of model will it be? A decision tree? Some other type of model? Some other parameters of the model type are specified too.
> نحدد ايش نوع نموذج تعلم الآلة اللي راح نستخدمه بالإضافة لتحديد الـ parameters

- **Fit**: Capture patterns from provided data. This is the heart of modeling.
> نشوف ايش الأنماط في البيانات اللي عندنا (وهو ما يسمى بتدريب النموذج)

- **Predict**: Just what it sounds like
> بعدها راح نختبر النموذج عشان يتنبأ بالقيم اللي نبغاها

- **Evaluate**: Determine how accurate the model's predictions are.
> أخيرًا راح نقيم دقة التنبؤات حق النموذج

In [ ]:
# راح نستخدم الـ decision tree model
from sklearn.tree import DecisionTreeRegressor

#### Define the Model

In [ ]:
# Define model
melbourne_model = DecisionTreeRegressor(random_state=1)

> Specifying a number for `random_state` ensures you get the same results in each run.

#### Fit the Model

In [ ]:
# Fit model
melbourne_model.fit(X,y)

DecisionTreeRegressor(random_state=1)

#### Make Predictions
هنا فقط للتمرين راح نتنبأ بقيمة نفس الـ houses اللي عندنا
> بس في الحقيقة نحن راح نتنبأ بقيم الـ new houses

In [ ]:
print(f"Making predictions for the following 5 houses:\n{X.head()}")

Making predictions for the following 5 houses:
   Rooms  Bathroom  Landsize  Lattitude  Longtitude
1      2       1.0     156.0   -37.8079    144.9934
2      3       2.0     134.0   -37.8093    144.9944
4      4       1.0     120.0   -37.8072    144.9941
6      3       2.0     245.0   -37.8024    144.9993
7      2       1.0     256.0   -37.8060    144.9954


In [ ]:
print(f'The predictions are:\n{melbourne_model.predict(X.head())}')

The predictions are:
[1035000. 1465000. 1600000. 1876000. 1636000.]


<a name='ex2'></a>
### Exercise: Your First Machine Learning Model

[**Housing Prices Competition for Kaggle Learn Users**](https://www.kaggle.com/competitions/home-data-for-ml-course/data)



#### Step 1: Specify Prediction Target
Select the **target variable**, which corresponds to the *sales price* `SalePrice`.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd

In [ ]:
iowa_file_path = '/content/drive/MyDrive/iMLND/Kaggle/Intro_to_ML/house_prices_iowa/iowa/train.csv'

In [ ]:
home_data = pd.read_csv(iowa_file_path)

In [ ]:
home_data.head()

,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.0,9550,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.0,14260,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,12,2008,WD,Normal,250000


نقدر نشوف اسماء الأعمدة بهذه الطريقتين التالية

In [ ]:
print(home_data.columns)

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

In [ ]:
home_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

الآن نختار العمود المطلوب عبر استخدام الـ dot notation

In [ ]:
home_data.SalePrice

0       208500
1       181500
2       223500
3       140000
4       250000
         ...  
1455    175000
1456    210000
1457    266500
1458    142125
1459    147500
Name: SalePrice, Length: 1460, dtype: int64

نخزن القيم في متغير y واللي راح يكون هو الـ prediction target حقنا

In [ ]:
y = home_data.SalePrice # Prediction Target

#### Step 2: Create X

الآن راح ننشئ DataFrame واللي فيها الـ predictive features

Since you want only some columns from the original data, you'll first create a list with the names of the columns you want in X.

You'll use just the following columns in the list:

- `LotArea`: Lot size in square feet
- `YearBuilt`: Original construction date
- `1stFlrSF`: First Floor square feet
- `2ndFlrSF`: Second floor square feet
- `FullBath`: Full bathrooms above grade
- `BedroomAbvGr`: Bedrooms above grade (does NOT include basement bedrooms)
- `TotRmsAbvGrd`: Total rooms above grade (does not include bathrooms)

In [ ]:
# Create the list of features
feature_names = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']

In [ ]:
X = home_data[feature_names] 

In [ ]:
X

,LotArea,YearBuilt,1stFlrSF,2ndFlrSF,FullBath,BedroomAbvGr,TotRmsAbvGrd
0,8450,2003,856,854,2,3,8
1,9600,1976,1262,0,2,3,6
2,11250,2001,920,866,2,3,6
3,9550,1915,961,756,1,3,7
4,14260,2000,1145,1053,2,4,9
...,...,...,...,...,...,...,...
1455,7917,1999,953,694,2,3,7
1456,13175,1978,2073,0,2,3,7
1457,9042,1941,1188,1152,2,4,9
1458,9717,1950,1078,0,1,2,5


##### Review Data


In [ ]:
# Description or statistics from X
X.describe()

,LotArea,YearBuilt,1stFlrSF,2ndFlrSF,FullBath,BedroomAbvGr,TotRmsAbvGrd
count,1460.000000,1460.000000,1460.000000,1460.000000,1460.000000,1460.000000,1460.000000
mean,10516.828082,1971.267808,1162.626712,346.992466,1.565068,2.866438,6.517808
std,9981.264932,30.202904,386.587738,436.528436,0.550916,0.815778,1.625393
min,1300.000000,1872.000000,334.000000,0.000000,0.000000,0.000000,2.000000
25%,7553.500000,1954.000000,882.000000,0.000000,1.000000,2.000000,5.000000
50%,9478.500000,1973.000000,1087.000000,0.000000,2.000000,3.000000,6.000000
75%,11601.500000,2000.000000,1391.250000,728.000000,2.000000,3.000000,7.000000
max,215245.000000,2010.000000,4692.000000,2065.000000,3.000000,8.000000,14.000000


In [ ]:
# The top few lines
X.head()

,LotArea,YearBuilt,1stFlrSF,2ndFlrSF,FullBath,BedroomAbvGr,TotRmsAbvGrd
0,8450,2003,856,854,2,3,8
1,9600,1976,1262,0,2,3,6
2,11250,2001,920,866,2,3,6
3,9550,1915,961,756,1,3,7
4,14260,2000,1145,1053,2,4,9


#### Step 3: Specify and Fit Model

In [ ]:
# Specify the model
from sklearn.tree import DecisionTreeRegressor

In [ ]:
iowa_model = DecisionTreeRegressor(random_state=1)

In [ ]:
# Fith the model
iowa_model.fit(X, y)

DecisionTreeRegressor(random_state=1)

#### Step 4: Make Predictions

In [ ]:
predictions = iowa_model.predict(X)

In [ ]:
predictions

array([208500., 181500., 223500., ..., 266500., 142125., 147500.])

##### Think About Your Results
نعمل مقارنة بين بعض الـ predictions والـ قيم الـ target

In [ ]:
y

0       208500
1       181500
2       223500
3       140000
4       250000
         ...  
1455    175000
1456    210000
1457    266500
1458    142125
1459    147500
Name: SalePrice, Length: 1460, dtype: int64

In [ ]:
iowa_model.predict(X.head())

array([208500., 181500., 223500., 140000., 250000.])

<a name='ch4'></a>
## **4- Model Validation**
Measure the performance of your model, so you can test and compare alternatives.\
https://www.kaggle.com/code/dansbecker/model-validation

هنا راح نقيس جودة نموذج تعلم الآلة واللي منها راح نعرف كيف نحسن من دقة النموذج حقنا

<a name='sec4-1'></a>
### 4-1 What is Model Validation

كيف نتحقق من جودة نموذج تعلم الآلة؟

أنسب مقياس لجودة النموذج هو الدقة التنبؤية predictive accuracy

يعني المطلوب معرفة هل التوقعات حق النموذج قريبة من اللي يصير في الواقع

Many people make a **huge mistake** when measuring predictive accuracy. They make predictions with their *training data* and compare those predictions to the target values in the *training data*. You'll see the problem with this approach and how to solve it.

توجد العديد من المقاييس لتلخيص جودة النموذج و واحدة منها تسمى Mean Absolute Error (MAE)



```
error = actual - predicted
```



> بس نحن راح ناخذ القيمة المطلقة لهذا الـ error


Absolute value of error
```
absolute error = |actual - predicted|
```



ومنها راح نجمع كل القيم حق الـ absolute error وبعدها راح ناخذ المتوسط حقهم

<a name='sec4-2'></a>
### 4-2  Calculate Mean Absolute Error (MAE)

In [ ]:
import pandas as pd

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
melbourne_data = pd.read_csv('/content/drive/MyDrive/iMLND/Kaggle/Intro_to_ML/melb_data.csv')

In [ ]:
# Filter rows with missing price values
filtered_melbourne_data = melbourne_data.dropna(axis = 0)

In [ ]:
# Choose target and features
y = filtered_melbourne_data.Price

melbourne_features = ['Rooms', 'Bathroom', 'Landsize', 'BuildingArea',
                      'YearBuilt', 'Lattitude', 'Longtitude']
X = filtered_melbourne_data[melbourne_features]

In [ ]:
from sklearn.tree import DecisionTreeRegressor

In [ ]:
# Define model
melbourne_model = DecisionTreeRegressor()

In [ ]:
# Fit model
melbourne_model.fit(X,y)

DecisionTreeRegressor()

In [ ]:
# Calculate the mean absolute error
from sklearn.metrics import mean_absolute_error

predicted_home_prices = melbourne_model.predict(X)

mean_absolute_error(y, predicted_home_prices)

434.71594577146544

> استخدمنا في الطريقة السابقة in-sample data وهذا غير مقبول في تحديد دقة النموذج

<a name='sec4-3'></a>
### 4-3 The Problem with "In-Sample" Scores

في الخطوة السابقة كما لاحظنا أن البيانات اللي استخدمناها في تدريب النموذج هي نفسه اللي قيمنا بها النموذج وهذا شي غير مقبول

السبب هو أن النموذج راح يطلع bias وراح تطلع لنا وكأن النتائج ودقة النموذج عالية

لكن لو جربنا النموذج على بيانات جديدة راح يفشل في التنبؤ بشكل سليم

لذلك يجب عند قياس اداء النموذج أن نستخدم بيانات جديدة لم يتم استخدامها أثناء بناء النموذج

وحدة من الطرق هي أننا نقسم البيانات إلى:
- training data
- **validation data** (لاختبار دقة النموذج)

في مكتبة scikit-learn توجد function باسم `train-test-split` لتقسم البيانات اللي لدينا:
-  **training data**: to fit the model
-  **validation data**: to calculate mean_absolute_error

هذا التقسم راح يتم على كلا الـ features والـ target

In [ ]:
from sklearn.model_selection import train_test_split

تقسيم البيانات راح يكون بشكل عشوائي بس نحن هنا راح نقيد عملية التقسيم ليكون التقسيم دائمًا نفسه عبر استخدام `random_state` argument

In [ ]:
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state = 0)

In [ ]:
# Define model
melbourne_model = DecisionTreeRegressor()

In [ ]:
# Fit model
melbourne_model.fit(train_X, train_y)

DecisionTreeRegressor()

In [ ]:
# Get predicted prices on validation data
val_predictions = melbourne_model.predict(val_X)

In [ ]:
mean_absolute_error(val_y, val_predictions)

267772.0058102001

> هنا نلاحظ الفرق بين القيمة اللي طلعناها سابقًا لمن استخدمنا in-sample data حيث كانت الـ MAE حوالي 434

> بس لمن استخدمنا الـ out-sample data طلعت لنا الـ MAE بحوالي 278000

In [ ]:
# the average home value in the validation data
val_y.mean()

1091516.642995481

In [ ]:
mean_absolute_error(val_y, val_predictions)/val_y.mean()

0.2453210471215039

> As a point of reference, *the average home value in the validation data is about 1.1 million dollars*. So **the error** in new data is **about a quarter** of the average home value.

There are many ways **to improve this model**, such as experimenting to find **better features** or **different model types**.

<a name='ex3'></a>
### Exercise: Model Validation

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
from sklearn.tree import DecisionTreeRegressor

iowa_file_path = '/content/drive/MyDrive/iMLND/Kaggle/Intro_to_ML/house_prices_iowa/iowa/train.csv'
home_data = pd.read_csv(iowa_file_path)

y = home_data.SalePrice
# Create the list of features
feature_names = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']
X = home_data[feature_names] 

# Specify the model
iowa_model = DecisionTreeRegressor()

# Fith the model
iowa_model.fit(X, y)

DecisionTreeRegressor()

In [ ]:
print(f"First in-sample prediction:\n{iowa_model.predict(X.head())}")

First in-sample prediction:
[208500. 181500. 223500. 140000. 250000.]


In [ ]:
print(f"Actual target values for homes:\n{y.head()}")

Actual target values for homes:
0    208500
1    181500
2    223500
3    140000
4    250000
Name: SalePrice, dtype: int64


#### Step 1: Split Your Data

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
train_X, val_X, train_y, val_y = train_test_split(X,y,random_state = 1)

#### Step 2: Specify and Fit the Model

In [ ]:
# Specify the model
iowa_model = DecisionTreeRegressor(random_state=1)

In [ ]:
# Fit iowa_model with the training data.
iowa_model.fit(train_X, train_y)

DecisionTreeRegressor(random_state=1)

#### Step 3: Make Predictions with Validation data

In [ ]:
# Predict with all validation observations
val_predictions = iowa_model.predict(val_X)

Inspect your predictions and actual values from validation data.

In [ ]:
val_predictions[:10]

array([186500., 184000., 130000.,  92000., 164500., 220000., 335000.,
       144152., 215000., 262000.])

In [ ]:
y[:10]

0    208500
1    181500
2    223500
3    140000
4    250000
5    143000
6    307000
7    200000
8    129900
9    118000
Name: SalePrice, dtype: int64

#### Step 4: Calculate the Mean Absolute Error in Validation Data

In [ ]:
from sklearn.metrics import mean_absolute_error

In [ ]:
val_mae = mean_absolute_error(val_y, val_predictions)

In [ ]:
val_mae

29652.931506849316

<a name='ch5'></a>
## **5- Underfitting and Overfitting**
Optimizing the Model 

Fine-tune your model for better performance.\
https://www.kaggle.com/code/dansbecker/underfitting-and-overfitting

عشان نحسن من دقة النموذج مثل decision tree

في هذا النوع نحتاج نتعمق ونتفرع أكثر عشان نحصل على prediction مناسبة

If a tree only had 1 split, it divides the data into 2 groups. If each group is split again, we would get 4 groups of houses. Splitting each of those again would create 8 groups. If we keep doubling the number of groups by adding more splits at each level, we'll have  $2^{10}$  groups of houses by the time we get to the 10th level. That's 1024 leaves.

زيادة التفرعات قد تقودنا للـ overfitting

يعني راح تكون جدًا دقيقة مع بيانات التدريب لكن لمن نستخدم النموذج مع بيانات جديدة راح يكون دقته جدًا ضعيفة

طيب لو قللنا التفرعات ايش ممكن يصير؟

لو كانت التفرعات قليلة جدًا هنا راح نقع في مشكلة ثانية وهي underfitting

هنا راح تكون نتائج التوقع بعيدة جدًا حتى عن بيانات التدريب وذلك بسبب أن النموذج فشل في تعلم الفروقات والأنماط المهمة من بيانات المعطى له لذلك أصبح أداءه ضعيف مع بيانات التدريب

<a name='sec5-1'></a>
### 5-1 Compare MAE Scores

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
melbourne_data = pd.read_csv('/content/drive/MyDrive/iMLND/Kaggle/Intro_to_ML/melb_data.csv')

# Filter rows with missing price values
filtered_melbourne_data = melbourne_data.dropna(axis = 0)

# Choose target and features
y = filtered_melbourne_data.Price

melbourne_features = ['Rooms', 'Bathroom', 'Landsize', 'BuildingArea',
                      'YearBuilt', 'Lattitude', 'Longtitude']
X = filtered_melbourne_data[melbourne_features]

#### Utility Function
راح تساعدنا عشان نقارن النتائج من MAE بعد تغيير الـ leaf nodes في كل مرة

In [ ]:
from sklearn.metrics import mean_absolute_error
from sklearn.tree import DecisionTreeRegressor

In [ ]:
def get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y):
    model = DecisionTreeRegressor(max_leaf_nodes = max_leaf_nodes, random_state = 0)
    model.fit(train_X, train_y)
    preds_val = model.predict(val_X)
    mae = mean_absolute_error(val_y, preds_val)

    return(mae)

In [ ]:
from sklearn.model_selection import train_test_split
train_X, val_X, train_y, val_y = train_test_split(X,y,random_state = 0)

#### Compare Accuracy of Models

In [ ]:
# مقارنة MAE بعد تغيير قيم الـ max_leaf_nodes
for max_leaf_nodes in [5,50,500,5000]:
    my_mae = get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y)

    print(f"Max leaf nodes: {max_leaf_nodes} \t\t Mean Absolute Error: {my_mae}")

Max leaf nodes: 5 		 Mean Absolute Error: 347380.33833344496
Max leaf nodes: 50 		 Mean Absolute Error: 258171.21202406782
Max leaf nodes: 500 		 Mean Absolute Error: 243495.96361790417
Max leaf nodes: 5000 		 Mean Absolute Error: 254983.64299548094


> نلاحظ أن كل ما زودنا قيمة الـ leaf nodes كل ما قلت قيمة الـ MAE\
لكن لمن تزيد الـ leaf nodes أكثر من اللازم راح ترجع قيمة الـ MAE وترتفع

Of the options listed, 500 is the optimal number of leaves.

<a name='ex4'></a>
### Exercise: Underfitting and Overfitting

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error

home_data = pd.read_csv('/content/drive/MyDrive/iMLND/Kaggle/Intro_to_ML/house_prices_iowa/iowa/train.csv')

y = home_data.SalePrice
# Create the list of features
feature_names = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']
X = home_data[feature_names] 

# Specify the model
iowa_model = DecisionTreeRegressor()

# Fith the model
iowa_model.fit(X, y)

# Split into validation & training data
train_X, val_X, train_y, val_y = train_test_split(X,y,random_state = 1)

# Specify the model
iowa_model = DecisionTreeRegressor(random_state=1)

# Fit iowa_model with the training data
iowa_model.fit(train_X, train_y)

# Predict with all validation observations
val_predictions = iowa_model.predict(val_X)

val_mae = mean_absolute_error(val_y, val_predictions)

In [ ]:
print(f'Validation MAE: {val_mae:,.0f}')

Validation MAE: 29,653


#### Step 1: Compare Different Tree Sizes

In [ ]:
def get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y):
    model = DecisionTreeRegressor(max_leaf_nodes=max_leaf_nodes, random_state=0)
    model.fit(train_X, train_y)
    preds_val = model.predict(val_X)
    mae = mean_absolute_error(val_y, preds_val)
    return(mae)

In [ ]:
candidate_max_leaf_nodes = [5, 25, 50, 100, 250, 500]

In [ ]:
for max_leaf_nodes in candidate_max_leaf_nodes:
    my_mae = get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y)
    print(f'Max leaf nodes: {max_leaf_nodes} \t\t Mean Absolute Error: {my_mae}')

Max leaf nodes: 5 		 Mean Absolute Error: 35044.51299744237
Max leaf nodes: 25 		 Mean Absolute Error: 29016.41319191076
Max leaf nodes: 50 		 Mean Absolute Error: 27405.930473214907
Max leaf nodes: 100 		 Mean Absolute Error: 27282.50803885739
Max leaf nodes: 250 		 Mean Absolute Error: 27893.822225701646
Max leaf nodes: 500 		 Mean Absolute Error: 29454.18598068598


In [ ]:
dict_mae={}
for max_leaf_nodes in candidate_max_leaf_nodes:
    my_mae = get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y)
    print(f'Max leaf nodes: {max_leaf_nodes} \t\t Mean Absolute Error: {my_mae}')
    dict_mae[max_leaf_nodes] = my_mae

print(dict_mae)

Max leaf nodes: 5 		 Mean Absolute Error: 35044.51299744237
Max leaf nodes: 25 		 Mean Absolute Error: 29016.41319191076
Max leaf nodes: 50 		 Mean Absolute Error: 27405.930473214907
Max leaf nodes: 100 		 Mean Absolute Error: 27282.50803885739
Max leaf nodes: 250 		 Mean Absolute Error: 27893.822225701646
Max leaf nodes: 500 		 Mean Absolute Error: 29454.18598068598
{5: 35044.51299744237, 25: 29016.41319191076, 50: 27405.930473214907, 100: 27282.50803885739, 250: 27893.822225701646, 500: 29454.18598068598}


In [ ]:
dict_mae.values()

dict_values([35044.51299744237, 29016.41319191076, 27405.930473214907, 27282.50803885739, 27893.822225701646, 29454.18598068598])

In [ ]:
min(dict_mae.values())

27282.50803885739

In [ ]:
dict_mae.get(5)

35044.51299744237

In [ ]:
for key, value in dict_mae.items():
    print(key,value)

5 35044.51299744237
25 29016.41319191076
50 27405.930473214907
100 27282.50803885739
250 27893.822225701646
500 29454.18598068598


In [ ]:
for key, value in dict_mae.items():
    if value == min(dict_mae.values()):
        print(key)

100


In [ ]:
dict_mae={}
for max_leaf_nodes in candidate_max_leaf_nodes:
    my_mae = get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y)
    print(f'Max leaf nodes: {max_leaf_nodes} \t\t Mean Absolute Error: {my_mae}')
    dict_mae[max_leaf_nodes] = my_mae

min_mae = min(dict_mae.values())
for key, value in dict_mae.items():
    if value == min(dict_mae.values()):
        print(key)

Max leaf nodes: 5 		 Mean Absolute Error: 35044.51299744237
Max leaf nodes: 25 		 Mean Absolute Error: 29016.41319191076
Max leaf nodes: 50 		 Mean Absolute Error: 27405.930473214907
Max leaf nodes: 100 		 Mean Absolute Error: 27282.50803885739
Max leaf nodes: 250 		 Mean Absolute Error: 27893.822225701646
Max leaf nodes: 500 		 Mean Absolute Error: 29454.18598068598
100


In [ ]:
def best_max_leaf_nodes(candidate_max_leaf_nodes):
    dict_mae={}
    for max_leaf_nodes in candidate_max_leaf_nodes:
        my_mae = get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y)
        dict_mae[max_leaf_nodes] = my_mae

    min_mae = min(dict_mae.values())
    for key, value in dict_mae.items():
        if value == min(dict_mae.values()):
            return key

In [ ]:
best_max_leaf_nodes([5, 25, 50, 100, 250, 500])

100

##### Solution from Kaggle

In [ ]:
scores = {leaf_size: get_mae(leaf_size, train_X, val_X, train_y, val_y) for leaf_size in candidate_max_leaf_nodes}
best_tree_size = min(scores, key=scores.get)

In [ ]:
best_tree_size

100

#### Step 2: Fit Model Using All Data

هنا راح نستخدم أفضل حجم للـ decision tree بس هذه المرة راح نجرب نحط كل الـ data اللي عندنا

In [ ]:
final_model = DecisionTreeRegressor(max_leaf_nodes = 100)
final_model.fit(X,y)

DecisionTreeRegressor(max_leaf_nodes=100)

<a name='ch6'></a>
## **6- Random Forests**
Using a more sophisticated machine learning algorithm.\
https://www.kaggle.com/code/dansbecker/random-forests

تعتبر الـ Random Forests Model أفضل من الـ single decision tree وبشكل عام دقة تنبؤها أفضل

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pandas as pd
melbourne_data = pd.read_csv('/content/drive/MyDrive/iMLND/Kaggle/Intro_to_ML/melb_data.csv')

# Filter rows with missing price values
filtered_melbourne_data = melbourne_data.dropna(axis = 0)

# Choose target and features
y = filtered_melbourne_data.Price

melbourne_features = ['Rooms', 'Bathroom', 'Landsize', 'BuildingArea',
                      'YearBuilt', 'Lattitude', 'Longtitude']
X = filtered_melbourne_data[melbourne_features]

# Split data into training & validation data
from sklearn.model_selection import train_test_split
train_X, val_X, train_y, val_y = train_test_split(X,y,random_state = 0)

In [3]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error

In [4]:
forest_model = RandomForestRegressor(random_state=1)
forest_model.fit(train_X, train_y)
melb_preds = forest_model.predict(val_X)
print(mean_absolute_error(val_y,melb_preds))

191669.7536453626


> نلاحظ أن قيمة MAE اللي حصلنا عليها أفضل من القيم السابقة اللي حصلنا عليها من الـ Decision Tree Models\
حيث كانت القيمة مع أفضل عدد من التفرعات بـ 500 تفرع هي حوالي `243495`

<a name='ex5'></a>
### Exercise: Random Forests

In [5]:
import pandas as pd
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error

home_data = pd.read_csv('/content/drive/MyDrive/iMLND/Kaggle/Intro_to_ML/house_prices_iowa/iowa/train.csv')

y = home_data.SalePrice
# Create the list of features
feature_names = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']
X = home_data[feature_names] 

# Specify the model
iowa_model = DecisionTreeRegressor()

# Fith the model
iowa_model.fit(X, y)

# Split into validation & training data
train_X, val_X, train_y, val_y = train_test_split(X,y,random_state = 1)

# Specify the model
iowa_model = DecisionTreeRegressor(random_state=1)

# Fit iowa_model with the training data
iowa_model.fit(train_X, train_y)

# Predict with all validation observations
val_predictions = iowa_model.predict(val_X)
val_mae = mean_absolute_error(val_y, val_predictions)
print(f'Validation MAE when not specifying max_leaf_nodes: {val_mae:,.0f}')


# Using best value for max_leaf_nodes
final_model = DecisionTreeRegressor(max_leaf_nodes = 100, random_state=1)
final_model.fit(train_X,train_y)
val_predictions = final_model.predict(val_X)
best_val_mae = mean_absolute_error(val_y, val_predictions)
print(f'Validation MAE for best value of max_leaf_nodes: {best_val_mae:,.0f}')

Validation MAE when not specifying max_leaf_nodes: 29,653
Validation MAE for best value of max_leaf_nodes: 27,283


#### Step 1: Use a Random Forest

In [6]:
from sklearn.ensemble import RandomForestRegressor

In [7]:
# Define the model
rf_model = RandomForestRegressor(random_state=1)

In [8]:
# Fit the model
rf_model.fit(train_X, train_y)

RandomForestRegressor(random_state=1)

In [9]:
# Calculate MAE of RF model
house_price_preds = rf_model.predict(val_X)
rf_val_mae = mean_absolute_error(val_y, house_price_preds)

print(f'Validation MAE for Random Forest: {rf_val_mae:,.0f}')

Validation MAE for Random Forest: 21,857


> زي ما نلاحظ أن قيمة MAE نزلت من `27,283` إلى `21,857`

<a name='ch7'></a>
## **7- Machine Learning Competitions**
Enter the world of machine learning competitions to keep improving and see your progress.\
https://www.kaggle.com/code/alexisbcook/machine-learning-competitions

In [10]:
import pandas as pd
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error

home_data = pd.read_csv('/content/drive/MyDrive/iMLND/Kaggle/Intro_to_ML/house_prices_iowa/iowa/train.csv')

y = home_data.SalePrice
# Create the list of features
feature_names = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']
X = home_data[feature_names] 

# Specify the model
iowa_model = DecisionTreeRegressor()

# Fith the model
iowa_model.fit(X, y)

# Split into validation & training data
train_X, val_X, train_y, val_y = train_test_split(X,y,random_state = 1)

# Define the model
rf_model = RandomForestRegressor(random_state=1)

# Fit the model
rf_model.fit(train_X, train_y)

# Calculate MAE of RF model
house_price_preds = rf_model.predict(val_X)
rf_val_mae = mean_absolute_error(val_y, house_price_preds)

print(f'Validation MAE for Random Forest: {rf_val_mae:,.0f}')

Validation MAE for Random Forest: 21,857


### Train a model for the competition

هنا راح نستخدم الملف كامل حق الـ training data اللي معطى من نفس الـ competition بدون تقسيم لأنه يوجد ملف خاص للـ test data

In [11]:
rf_model_on_full_data = RandomForestRegressor(random_state=1)
rf_model_on_full_data.fit(X,y)

RandomForestRegressor(random_state=1)

In [12]:
test_data = pd.read_csv('/content/drive/MyDrive/iMLND/Kaggle/Intro_to_ML/house_prices_iowa/iowa/test.csv')

In [14]:
test_X = test_data[feature_names]
test_preds = rf_model.predict(test_X)

In [15]:
# output = pd.DataFrame({'Id': test_data.Id, 'SalePrice': test_preds})
# output.to_csv('submission.csv', index=False)